stackrox/stackrox
# ROX-23946: Factor out email template from vulnerability reporting

## Description

Prerequisite to factor out delivery destinations from Vulnerability Reports to reuse in Compliance Reports.

### Changes

1. Create EmailTemplate.utils.ts
    * Move maximum lengths from useReportFormValues.tsx file and leave what is specific to vulnerability reports.
    * Move reusable validations from emailTemplateFormUtils.ts file and leave what is specific to vulnerability reports.
    * Rename variables for consistency with `NotifierConfiguration` type.
2. Create EmailTemplateForm.tsx
    * Rename variables for consistency with `NotifierConfiguration` type.
3. Move and rename EmailTemplateModal.tsx
    * Replace `reportParameters` with optional `renderTemplatePreview` prop.
    * Rename variables for consistency with `NotifierConfiguration` type.
    * Factor out form.
    * Conditionally render tabs.
4. Edit DeliveryDestinationsForm.tsx
    * Import from `Components/EmailTemplate` folder.
    * Factor out `renderTemplatePreview` which is specific to vulnerability reports.
5. Edit emailTemplateFormUtils.ts
    * Move reusable validations.
6. Edit useReportFormValues.tsx
    * Move reusable lengths and validations.

## Checklist

## Testing Performed

1. `yarn lint` in ui
2. `yarn build` in ui and then compute branch - master for the following
    * `wc build/static/js/*.js`
        main.js 0 = 4633558 - 4633558
        total 353 = 11601343 - 11600990
    * `ls -al build/static/js/*.js | wc`
        files 0 = 173 - 173
3. `yarn start` in ui

### Manual testing

Prerequisite: create at least one collection.

Create e-mail notifiers on the fly.

1. Create 3 reports with various combinations of defaut or custom destinations or template.
    * Verify payload and response of POST /v1/notifiers request.
    * Verify review details.
    * Verify payload and response of POST /v2/reports/configurations request.
2. Edit the first report and ditto items 2 and 3 above.

EmailTemplateModal **Edit** tab
![EmailTemplateModal_1_Edit](https://github.com/stackrox/stackrox/assets/11862657/f45d99f9-23be-4505-8578-7bd22c8a566d)

EmailTemplateModal **Preview** tab and **CVEs found** toggle
![EmailTemplateModal_1_Preview_CVEs_found](https://github.com/stackrox/stackrox/assets/11862657/01b2c3a2-6cfb-49ec-93a3-8f1ab02c59d8)

EmailTemplateModal **Preview** tab and **CVEs not found** toggle
![EmailTemplateModal_1_Preview_CVEs_not_found](https://github.com/stackrox/stackrox/assets/11862657/2ebdaa53-6226-49ae-8c64-32eb22e958a8)

EmailTemplateModal with EmailTemplateForm only (that is, without `renderTemplatePreview` prop, which compliance reports probably not have)
![EmailTemplateModal_with_EmailTemplateForm_only](https://github.com/stackrox/stackrox/assets/11862657/edcd00fc-e9fe-4a08-a59a-1086cea02fc0)

DeliveryDestinationsDetails
![DeliveryDestinationsDetails](https://github.com/stackrox/stackrox/assets/11862657/32cba022-e1b2-411f-b15c-ab1fd2618632)

EmailTemplatePreview
![EmailTemplatePreview](https://github.com/stackrox/stackrox/assets/11862657/a689a6e8-6297-4ca8-ab1e-a443b19c27f8)

### Integration testing

Uh, oh! No tests. Yet.
